toto/in_toto_record.py

```diff
@@ -0,0 +1,221 @@
+  """
+  <Purpose>
+    Finishes creating link metadata for a multi-part in-toto step.
+    Loads signing key and unfinished link metadata file from disk, verifies
+    that the file was signed with the key, records products, updates unfinished
+    Link object (products and signature), removes unfinished link file from and
+    stores new link file to disk.
+
+  <Arguments>
+    step_name:
+            A unique name to relate link metadata with a step defined in the
+            layout.
+    key_path:
+            Path to private key to verify unfinished link metadata and
+            sign finished link metadata (PEM).
+    product_list:
+            List of file or directory paths that should be recorded as products.
+
+  <Exceptions>
+    Exception if signature fails
+    IOException if file ".<step_name>.link-unfinished" can't be read
+
+  <Side Effects>
+    Loads private key from disk
+    Writes link file to disk "<step_name>.link"
+    Remvoes unfinished link file ".<step_name>.link-unfinished" from disk
+    Calls sys.exit(1) if an exception is raised
+
+  <Returns>
+    Link object
+
+  """
+
+  unfinished_fn = "." + str(step_name) + ".link-unfinished"
+
+  try:
+    log.doing("load link signing key...")
+    key = toto.util.prompt_import_rsa_key_from_file(key_path)
+    keydict = {key["keyid"] : key}
+
+    # There must be a file .<step_name>.link-unfinished in the current dir
+    log.doing("load unfinished link file...")
+    link = Link.read_from_file(unfinished_fn)
+    # The unfinished link file must have been signed by the same key
+
+    log.doing("verify unfinished link signature...")
+    link.verify_signatures(keydict)
+
+    log.doing("record products...")
+    link.products = runlib.record_artifacts_as_dict(product_list)
+
+    log.doing("update signature...")
+    link.signatures = []
+    link.sign(key)
+
+    log.doing("store metadata to file...")
+    link.dump()
+
+    log.doing("remove unfinished file...")
+    os.remove(unfinished_fn)
+
+    return link
+
+  except Exception, e:
+    log.error("in stop record - %s" % e)
+    sys.exit(1)
+
+
+
+def main():
+  parser = argparse.ArgumentParser(
+      description="Starts or stops link metadata recording")
+
+  subparsers = parser.add_subparsers(dest="command")
+
+  subparser_start = subparsers.add_parser('start')
+  subparser_stop = subparsers.add_parser('stop')
+
+  # Whitespace padding to align with program name
+  lpad = (len(parser.prog) + 1) * " "
+  parser.usage = ("\n"
+      "%(prog)s  --step-name <unique step name>\n{0}"
+               " --key <functionary private key path>\n"
+      "Commands:\n{0}"
+               "start [--materials <filepath>[ <filepath> ...]]\n{0}"
+               "stop  [--products <filepath>[ <filepath> ...]]\n"
+               .format(lpad))
+
+  in_toto_args = parser.add_argument_group("Toto options")
+  # FIXME: Name has to be unique!!! Where will we check this?
+  # FIXME: Do we limit the allowed characters for the name?
```

Comment:
Good point, this is not on the specification, so I need to consider it. I wouldn't worry about this in here though.